### PR TITLE
feat: allow selection of iOS sim with `--udid`

### PR DESCRIPTION
**Pull request by alloy**
Summary:
---------

This makes it possible to be very specific about what simulator to use in the same way you’d interact with tooling such as `simctl`.

Fixes https://github.com/react-native-community/cli/issues/1078

Test Plan:
----------

Given a device list like:

```
/t/RNTestProject » xcrun instruments -s
Known Devices:
Eloy’s iMac [E163381F-C94E-5B5E-922E-DBB59B685FA6]
Eloy’s iPhone (13.3.1) [00008030-000D19512210802E]
iPhone 11 (13.3) [60E7F277-DC13-46B8-ADFB-7C5C9BC01274] (Simulator)
iPhone 8 (12.4) [0F518318-4C9B-49BA-89B9-1D0E325681BB] (Simulator)
```

### Default

```
/t/RNTestProject » ./node_modules/.bin/react-native run-ios
info Found Xcode workspace "RNTestProject.xcworkspace"
info Launching iPhone 11 (iOS 13.3)
info Building (using "xcodebuild -workspace RNTestProject.xcworkspace -configuration Debug -scheme RNTestProject -destination id=60E7F277-DC13-46B8-ADFB-7C5C9BC01274")
```

### By UDID

```
/t/RNTestProject » ./node_modules/.bin/react-native run-ios --udid 0F518318-4C9B-49BA-89B9-1D0E325681BB
info Found Xcode workspace "RNTestProject.xcworkspace"
info Launching iPhone 8 (iOS 12.4)
info Building (using "xcodebuild -workspace RNTestProject.xcworkspace -configuration Debug -scheme RNTestProject -destination id=0F518318-4C9B-49BA-89B9-1D0E325681BB")
```

### Default without iOS device connected

```
/t/RNTestProject » ./node_modules/.bin/react-native run-ios --device
info Found Xcode workspace "RN062rc3.xcworkspace"
error No iOS devices connected.
```

### Default with iOS device connected

```
/t/RNTestProject » ./node_modules/.bin/react-native run-ios --device
info Found Xcode workspace "RNTestProject.xcworkspace"
info Using first available device named "Eloy’s iPhone" due to lack of name supplied.
info Building (using "xcodebuild -workspace RNTestProject.xcworkspace -configuration Debug -scheme RNTestProject -destination id=00008030-000D19512210802E")
```

### By UDID

```
/t/RNTestProject » ./node_modules/.bin/react-native run-ios --udid 00008030-000D19512210802E
info Found Xcode workspace "RNTestProject.xcworkspace"
info Building (using "xcodebuild -workspace RNTestProject.xcworkspace -configuration Debug -scheme RNTestProject -destination id=00008030-000D19512210802E")
```

### By name

```
/t/RNTestProject » ./node_modules/.bin/react-native run-ios --device 'Eloy’s iMac'
info Found Xcode workspace "RN062rc3.xcworkspace"
info Building (using "xcodebuild -workspace RN062rc3.xcworkspace -configuration Debug -scheme RN062rc3 -destination id=E163381F-C94E-5B5E-922E-DBB59B685FA6")
```

### By UDID

```
/t/RNTestProject » ./node_modules/.bin/react-native run-ios --udid E163381F-C94E-5B5E-922E-DBB59B685FA6
info Found Xcode workspace "RN062rc3.xcworkspace"
info Building (using "xcodebuild -workspace RN062rc3.xcworkspace -configuration Debug -scheme RN062rc3 -destination id=E163381F-C94E-5B5E-922E-DBB59B685FA6")
```

### CODE REVIEW

File: packages/platform-ios/src/commands/runIOS/__tests__/parseIOSDevicesList.test.ts [hunk 1]
```diff
@@ -29,14 +29,29 @@ describe('parseIOSDevicesList', () => {
     );
 
     expect(devices).toEqual([
-      {
-        name: 'Maxs MacBook Pro',
-        udid: '11111111-1111-1111-1111-111111111111',
```
Comment (thymikee) [inline packages/platform-ios/src/commands/runIOS/__tests__/parseIOSDevicesList.test.ts]: why was this removed? we support running on host Mac for Catalyst

File: packages/platform-ios/src/commands/runIOS/parseIOSDevicesList.ts [hunk 2]
```diff
@@ -8,36 +8,24 @@
 import {Device} from '../../types';
 
 /**
- * Parses the output of `xcrun simctl list devices` command
+ * Parses the output of `xcrun instruments -s` command
```
Comment (thymikee) [inline packages/platform-ios/src/commands/runIOS/parseIOSDevicesList.ts]: good catch

File: packages/platform-ios/src/commands/runIOS/parseIOSDevicesList.ts [hunk 3]
```diff
@@ -8,36 +8,24 @@
...
-      const name = device[1];
-      const version = device[2];
-      const udid = device[3];
-      devices.push({udid, name, version});
+  text.split('\n').forEach(line => {
+    const device = line.match(
+      /(.*?) \(([0-9\.]+)\) \[([0-9A-F-]+)\]( \(Simulator\))?/,
+    );
+    if (device) {
+      const {[1]: name, [2]: version, [3]: udid, [4]: isSimulator} = device;
```
Comment (thymikee) [inline packages/platform-ios/src/commands/runIOS/parseIOSDevicesList.ts]: Wouldn't array destructuring be a bit more obvious? What do you think?
```suggestion
      const [, name, version, udid, isSimulator] = device;
```

File: packages/platform-ios/src/commands/runIOS/index.ts [hunk 4]
```diff
@@ -71,43 +71,51 @@ function runIOS(_: Array<string>, ctx: Config, args: FlagsT) {
...
+    }
+    if (device.isSimulator) {
+      return runOnSimulator(xcodeProject, scheme, args);
+    } else {
+      return runOnDevice(device, scheme, xcodeProject, args);
+    }
+  } else {
+    const physicalDevices = devices.filter(d => !d.isSimulator);
+    if (physicalDevices.length === 0) {
+      return logger.error('No iOS devices connected.');
```
Comment (thymikee) [inline packages/platform-ios/src/commands/runIOS/index.ts]: What about host mac? Isn't it always displayed by `xcrun instruments -s`?

File: packages/platform-ios/src/commands/runIOS/__tests__/parseIOSDevicesList.test.ts (hunk 1 above)
Comment (alloy) [reply packages/platform-ios/src/commands/runIOS/__tests__/parseIOSDevicesList.test.ts]: Ohhh, that’s good to know, thanks! I guess the function name made me think it was pure iOS.

File: packages/platform-ios/src/commands/runIOS/index.ts (hunk 4 above)
Comment (alloy) [reply packages/platform-ios/src/commands/runIOS/index.ts]: Not after the incorrect change I made above :) I thought the above was unnecessarily returning the Mac only to remove it here.

File: packages/platform-ios/src/commands/runIOS/__tests__/parseIOSDevicesList.test.ts (hunk 1 above)
Comment (thymikee) [reply packages/platform-ios/src/commands/runIOS/__tests__/parseIOSDevicesList.test.ts]: I get you, we just recently added this :D

**Comment by alloy** [comment]
Alright, added Catalyst support back and made it a bit more clear that that’s a supported target.

**Comment by thymikee** [comment]
@grabbou mind having a look?

File: packages/platform-ios/src/commands/runIOS/parseIOSDevicesList.ts [hunk 5]
```diff
@@ -8,36 +8,36 @@
...
-      if (myMac) {
-        const name = myMac[1];
-        const udid = myMac[2];
-        devices.push({
-          udid,
-          name,
-        });
+  text.split('\n').forEach(line => {
+    const device = line.match(
+      /(.*?) (\(([0-9\.]+)\) )?\[([0-9A-F-]+)\]( \(Simulator\))?/i,
```
Comment (thymikee) [inline packages/platform-ios/src/commands/runIOS/parseIOSDevicesList.ts]: Added lowercase flag, because some UUIDs may be lower-cased (discovered while local testing)

Comment (alloy) [reply packages/platform-ios/src/commands/runIOS/parseIOSDevicesList.ts]: Oh, good catch; thanks! Was it an iPhone?

Comment (thymikee) [reply packages/platform-ios/src/commands/runIOS/parseIOSDevicesList.ts]: yup, iPhone 8 to be exact

**Comment by amma-dev** [comment]
just update the cli-platform-ios inside @react-native-community/
try this
npm i @react-native-community/cli-platform-ios

hope it works
